### ISSUE TITLE
Metric catalogs, dynamic per task configurations

### ISSUE THREAD

**Issue by evilezh** [intelsdi-x/snap#1351]
I was trying to figure out - how to configure collector plugins per task. And point is - it is not possible within current design. 
Many plugins (most ?), read their possible metrics, when called GetMetricTypes, - **read dynamically** (scan something). That means - you must have configuration before you load plugin. Only way to do so - is to put config in GLOBAL configuration.
Which means:
- plugins are not independent !!!
- dynamic metric names is not supported
- possible fails if some metrics is not reported or missing

And main question (we can make survey). How many people have written automation around `METRIC CATALOG`. I guess that is one of features about nice to have, but who needs/uses it really ? If it is 2 from 100000 - is it worth then ?

Few ideas:

Tasks: 
1. Define which collector plugins you want to use, and their configurations. Same way like publisher plugins.
2. Implement regex filter per task to filter out data you want or don't want. (`include_regexp`,`exclude_regexp`)
Imagine something like:
```json
{  
   "collect":{  
      "metrics":[  
         {  
            "plugin_name":"mesos",
            "config":{  
               "agent":{  
                  "address":"http://localhost:5050"
               }
            },
            "filters":{  
               "include":".*",
               "exclude":"/intel/mesos/agent/tasks/.*test_instance-.*"
            }
         },
         {  
            "plugin_name":"cpu"
         }
      ],
      "process":null,
      "publish":[  
         {  
            "plugin_name":"kafka",
            "config":{  
               "topic":"snap-metrics-test-1",
               "brokers":"kafka.service:9200"
            }
         }
      ]
   }
}
```
If you still need metric catalog:
METRIC CATALOG = TASK which runes once without filters and returns no data (that is how currently most plugins work - they query system for available stats, and return namespace e.g. path only).

Plugin manager would just contain possible plugins+version available for use in tasks.

I think those are quite a simple changes which would unlock new possibilities and open way for better use. Also - it would keep somehow backward compatible "features".

**Comment by kindermoumoute** [comment]
> Many plugins (most ?), read their possible metrics, when called GetMetricTypes, - read dynamically (scan something). That means - you must have configuration before you load plugin. Only way to do so - is to put config in GLOBAL configuration.

This is how you pass a configuration to a collector when you start a task:
```json
{
    "version": 1,
    "schedule": {
        "type": "simple",
        "interval": "2s"
    },
    "max-failures": 2,
    "workflow": {
        "collect": {
            "metrics": {
                "|intel|mock|/foo=㊽": {},
                "|intel|mock|/bar⽔": {}
            },
            "config": {
                "/intel/mock": {
                    "user": "root",
                    "password": "secret"
                }
            },
            "process": [
                {
                    "plugin_name": "statistics",
                    "publish": [
                        {
                            "plugin_name": "file",                            
                            "config": {
                                "file": "/tmp/mock-stats.json"
                            }
                        }
                    ]
                }
            ]
        }
    }
}
```

Is it what you are trying to do?

**Comment by evilezh** [comment]
Not exactly. Plugins needs address/use/password before plugin manager loads in order to query available metrics names.

**Comment by kindermoumoute** [comment]
After talking with @evilezh on Slack, I think we can sum up 2 proposals from this issue:

1. Store **plugin configuration inside the plugin** instead of Snap itself, so we don't pass it every time through gRPC. We could imagine passing the config when we start the plugin.

2. A lot of collectors require a config in order to update their metric catalog. In order to improve user experience, we could **validate namespaces on the plugin side when we start a task**. This would potentially make metric catalog obsolete.

**Comment by andrzej-k** [comment]
Ad 1.
Some time ago we had a discussion how to deliver configuration to plugin, for example in cases when returning metric catalog requires some type of discovery (which in turn requires authentication for example). Today we have two ways of delivering config to plugin: a) global config and b) plugin config. The problem with plugin config is that it's a part of task manifest, so it cannot be used for metrics discovery mechanism. If I understand correctly you are proposing to create a new config (per plugin) which is not tied to task manifest - **is that right?**
Since using global config to provide per plugin configuration might be inconvenient, we are trying to make metric catalog static, meaning that if possible plugin should return list of metrics  without any need of additional configuration being provided.

Ad 2.
How is this related to point 1, does it depend on 1? 
So you proposing that metric catalog will be available after loading a plugin, but will be updated after task is started? Meaning that some new metrics will appear and some disappear on metric catalog after task is started?

**Comment by evilezh** [comment]
Here is main topics:
1. You have n-plugins available on system (binaries).
2. You want to load binaries with specific configuration.
3. **TASK** defines what you want to do.
4. Publisher plugins already have nice way to define configuration.
5. You might want to load same plugin, but with different configuration. In such case metrics might be slightly different.
6. available metrics might change over time
7.  someone told that metrics catalog can be used in automation (does anyone uses yet).

Implementation/explanation:
So - when we upload plugin, we would just make it available on system  (key is "plugin name" + version). Later in task - i can say, please use 
- input plugin "P" version "V" and configuration is "C"
- use processing plugin "PP", version "PV" and configuration "PC" 
- output use please plugin "OP", version "OV" and configuration "OC". 
As yo see - it follows same rules, same pattern.

**Namespaces.** Collectors define namespace /asd/asd/asd. How it is different from `"PLUGIN_
NAME":"VERSION"`. It is just some ID used for plugin.
That is important - because now you use namespace to determine which plugin to use for collection and to which plugin forward request to collect metrics, which results in following json:
```json
"collect": {
      "metrics": {
        "/intel/psutil/load/load1": {},
        "/intel/psutil/load/load15": {}
      },
```
As i suggested in my previous post - use plugin name + version to tell which plugin you want to run. 
```json
"metrics":[  
         {  
            "plugin_name":"mesos",
            "version":3,
            "config":{  
               "agent":{  
                  "address":"http://localhost:5050"
               }
            },
```
Then you can nicely attach configuration to exact plugin as well. Using namespaces it is not possible in nice way. And in case of filtering use regex - which is much more powerful way to filter what you want to collect (you can even reference patterns for dynamic parts).
**FILTERING** - how it is done currently: plugin runs, then collect all metrics, then apply filters /asd/*/asd/. Either way - it collect ALL data and then applies filters. So - why not regex then ? 

Metric catalog - you can add new endpoint POST /snapshot/<plugin-name>/<plugin-version> with optional config{} in body, which would return instantly available metrics. That still allows automation and is much more reliable and precise and flexible than configuration in GLOBAL.
